core/lib/dal/src/connection/processor.rs

```diff
@@ -0,0 +1,222 @@
+use std::{
+    collections::HashMap,
+    fmt,
+    panic::Location,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Mutex,
+    },
+    time::{Duration, Instant, SystemTime},
+};
+
+use sqlx::{pool::PoolConnection, types::chrono, Connection, PgConnection, Postgres, Transaction};
+
+use crate::metrics::CONNECTION_METRICS;
+
+/// Tags that can be associated with a connection.
+#[derive(Debug, Clone, Copy)]
+pub(super) struct StorageProcessorTags {
+    pub requester: &'static str,
+    pub location: &'static Location<'static>,
+}
+
+impl fmt::Display for StorageProcessorTags {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            formatter,
+            "requested by `{}` at {}:{}",
+            self.requester,
+            self.location.file(),
+            self.location.line()
+        )
+    }
+}
+
+struct TracedConnectionInfo {
+    tags: Option<StorageProcessorTags>,
+    created_at: Instant,
+}
+
+impl fmt::Debug for TracedConnectionInfo {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let timestamp: chrono::DateTime<chrono::Utc> =
+            (SystemTime::now() - self.created_at.elapsed()).into();
+        let tags_display: &dyn fmt::Display = match &self.tags {
+            Some(tags) => tags,
+            None => &"not tagged",
+        };
+        write!(formatter, "[{timestamp} - {tags_display}]")
+    }
+}
+
+/// Traced active connections for a connection pool.
+#[derive(Default)]
+pub(super) struct TracedConnections {
+    connections: Mutex<HashMap<usize, TracedConnectionInfo>>,
+    next_id: AtomicUsize,
```

Comment:
Maybe, there's a better way to track connections? Since `HashMap` entries are dropped when connections are dropped, the `HashMap` shouldn't grow unboundedly.